### ISSUE TITLE
new KEP for "kube proxy lib" 

### ISSUE THREAD

**Issue by jayunit100** [kubernetes-sigs/kpng#394]
TLDR We need to decide on wether or not staging/ is really going to be the right place for this code to live, to me, thats a function of "will core k8s depend on it". 

# Details

We have 2 options as far as what functionality we carry in kpng: Redundant k/k/proxy/... functionality, or, let k/k/proxy live on (until a later date when it wants to be coupled).  I think we need to  pick a horse: A hybrid model where we jam a bunch of code to in-tree ... seems like an anti-pattern in many ways, esp, per @aojea 's comment that folks are trying to move things *out of tree* for K8s.

## Option 0: K/k is totally coupled to KPNG

- kubernetes-sigs/kpng becomes a historical project
- k/k/staging/kpng has kpng as a library, implementing https://docs.google.com/presentation/d/1Y-tZ4fFC9L2NvtBeiIXD1MiJ0ieg_zg4Q6SlFmIax8w/edit?hl=en&resourcekey=0-SFhIGTpnJT5fo6ZSzQC57g#slide=id.g16976fedf03_0_201 
- k/k/proxy/iptables ... and other **in tree** proxies use k/k/staging/kpng

## Option 1: KPNG is vendorable, standalone proxy swiss army knife

- kubernetes-sigs/kpng is a vendorable, standalone library w/ example backends, vendors Only client go and nothing else from k8s. 
  - kernelspace windows (fullstate)
  - nft (fullstate)
  - ebpf (fullstate)
  - userspace windows and linux (filterReset decoder)
- kubernetes/kuberneetes/pkg/proxy contains:
  - iptables
  - ipvs
  - winkernel (fullstate, which vendors KPNG, from some place)

Lets use this issue to make a decision and weigh in on what we should do.  I dont have a strong opinion for one or the other, but...

## I actually (for once) do have a strong opinion that, we should probably not, be jamming unused code into ./staging/ which, is already complex to reason about and import....

if for no other reason then, that we dont want to waste sig-arch's time :)

**Comment by jayunit100** [comment]
Just posted this on the mailing list, but not sure it went through: talked w/ @rajaskakodkar and he talked with @palnabarun  and @nikhita .... 


So, circling back a few weeks later, and hearing from rajas, nikhita, naburun, and others....
- i think in KPNG we have, a new type of CAR 
- but what sig-net is asking for, is a BATTERY

Antonio's original comment deserves imo another look: 

"I still don't understand what is the need to have the code in the kubernetes repo"  

This whole thread can probably be coursely summarized as "Lets just build a battery, not a rebuild the car (yet)"

If thats the case, why not just let folks define that battery.... ? i.e. something like: 

- create a library , in a place like kubernetes-sigs/kube-proxy-lib or whatever... put the original kube proxy maintainers as the OWNERS.md on it 
- leverage parts from KPNG that are usable, i.e. in https://github.com/kubernetes-sigs/kpng/tree/master/server/jobs/kube2store  
- leverage  stoyocos' deltas in  https://github.com/kubernetes-sigs/kpng/pull/389 as a "first pass" at the minimal footprint of a kube-proxy-lib  
- migrate things like dans doc "[So you wanna build a Service Proxy ](https://github.com/kubernetes-sigs/kpng/blob/master/doc/service-proxy.md)"  into that repo , and update it to leverage kube-proxy-lib
- publish and release the lib w/ a ref impl (eBPF or userspace or a "dummy svc proxy" that prints services/endpoints/topologies... or something else that makes it self-verifyable and makes it concrete/easy to learn about).
- test the lib on specific k8s versions, and version it w/ support for different k8s versions and so on
- write  a very small KEP, based on   [tims slides](http://tinyurl.com/2v9fbucd) that describes this, and references lessons learned trying to make KPNG a replacement and, try to make it unopinionated, so that it can move forward w/ less friction and cognitive load.
